README.md
# Sharetribe Flex Template for Web: time-based process 

[![CircleCI](https://circleci.com/gh/sharetribe/ftw-hourly.svg?style=svg)](https://circleci.com/gh/sharetribe/ftw-hourly)

> **Note** If you are using day-based booking process (i.e. daily or nightly bookings), you should
> choose the original [Flex Template for Web](https://github.com/sharetribe/ftw-daily/) instead.

This is a template web application for a Sharetribe Flex marketplace using **time-based booking
process**. This repository is a fork from the existing
[FTW repository](https://github.com/sharetribe/ftw-hourly) (which is using nightly booking process).
So, if you have started working with the original FTW repository and you want to change to
FTW-hourly, you can
[change the 'upstream' remote repository](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/configuring-a-remote-for-a-fork).
You can read more about time-based process from the related
[Flex Docs article](https://www.sharetribe.com/docs/background/time-based-template)

## Quick start

> **Note:** this template assumes that you are using
> [time-based transaction process](https://www.sharetribe.com/docs/guides/how-to-take-time-based-bookings-into-use/)
> with alias `flex-hourly-default-process/release-1`, if you are using something else, you should
> change the alias name from
> [config.js](https://github.com/sharetribe/ftw-hourly/blob/master/src/config.js#L35) and
> [util/transaction.js](https://github.com/sharetribe/ftw-hourly/blob/master/src/util/transaction.js#L111)

If you just want to get the app running quickly to test it out, first install
[Node.js](https://nodejs.org/) and [Yarn](https://yarnpkg.com/), and follow along:

```sh
git clone git@github.com:sharetribe/ftw-hourly.git             # clone this repository
cd ftw-hourly/                                                 # change to the cloned directory
yarn install                                                   # install dependencies
yarn run config                                                # add the mandatory env vars to your local config
yarn run dev                                                   # start the dev server, this will open a browser in localhost:3000
```

You can also follow along the
[Getting started with FTW](https://www.sharetribe.com/docs/tutorials/getting-started-with-ftw/)
tutorial and read the background article about
[Time-based template](https://www.sharetribe.com/docs/background/time-based-template/) from the
[Flex Docs website](https://www.sharetribe.com/docs/).

For more information of the configuration, see the
[FTW Environment configuration variables](https://www.sharetribe.com/docs/references/ftw-env/)
reference in Flex Docs.

> **Note:** If you want to build your own Flex marketplace on top of the template, you should fork
> the repository instead of cloning it. See the
> [How to Customize FTW](https://www.sharetribe.com/docs/guides/how-to-customize-ftw/) guide in Flex
> Docs.

## Getting started with your own customization

If you want to build your own Flex marketplace by customizing the template application, see the
[How to Customize FTW](https://www.sharetribe.com/docs/guides/how-to-customize-ftw/) guide in Flex
Docs.

## Deploying to Heroku

> **Note:** Remember to fork the repository before deploying the application. Connecting your own
> Github repository to Heroku will make manual deploys easier.

See the
[How to deploy FTW to production](https://www.sharetribe.com/docs/guides/how-to-deploy-ftw-to-production/)
guide in Flex Docs for more information.

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

## Documentation

See the Flex Docs site: https://www.sharetribe.com/docs/

See also the [docs/](docs/) directory for some additional internal documentation.

## Get help – join Sharetribe Flex Developer Slack channel

If you have any questions about development, the best place to ask them is the Flex Developer Slack
channel at https://www.sharetribe.com/flex-slack

## License

This project is licensed under the terms of the Apache-2.0 license.

See [LICENSE](LICENSE)


CHANGELOG.md
# Change Log

We are not following semantic versioning in this template app since any change could potentially be
a breaking change for forked customization projects. We are still experimenting with what is a good
way to update this template, but currently, we follow a pattern:

- Major version change (v**X**.0.0): Changes to several pages and other components. Consider
  implementing this without merging upstream (we'll provide instructions).
- Minor version change (v0.**X**.0): New features and changes to a single page. These are likely to
  cause conflicts.
- Patch (v0.0.**X**): Bug fixes and small changes to components.

Note: Previous release tags (before v4.0.0) are coming from upstream:
https://github.com/sharetribe/flex-template-web/

---

## Upcoming version 2020-XX-XX

- [change] Request custom image variants for avatar [#110](https://github.com/sharetribe/ftw-hourly/pull/110)

## [v8.1.1] 2020-07-21

- [fix] Fix config script for NodeJS v14.5.0
  [#1327](https://github.com/sharetribe/ftw-daily/pull/1327)

[v8.1.1]: https://github.com/sharetribe/ftw-hourly/compare/v8.1.0...v8.1.1

## [v8.1.0] 2020-07-01

- [change] Change the transaction process name to be `flex-hourly-default-process/release-1`
  [#104](https://github.com/sharetribe/ftw-hourly/pull/104)

### Updates from upstream (FTW-daily v6.1.0)

- [fix] MainPanel: search filter bug. Address and bounds are handled outside of MainPanel, URL
  params should be trusted instead of values stored to state.
  [#1320](https://github.com/sharetribe/ftw-daily/pull/1320)
- [fix] small typo. [#1319](https://github.com/sharetribe/ftw-daily/pull/1319)
- [fix] Fix typo (which is copy-pasted in 4 files).
  [#1318](https://github.com/sharetribe/ftw-daily/pull/1318)
- [add] Update French translation file (Spanish and German translations have still missing keys).
  [#1316](https://github.com/sharetribe/ftw-daily/pull/1316)
- [fix] Sync bookingUnitType variables and update comments. Client app's API (proxy) server needs to
  know about unit type. [#1317](https://github.com/sharetribe/ftw-daily/pull/1317)

[v8.1.0]: https://github.com/sharetribe/ftw-hourly/compare/v8.0.0...v8.1.0

## [v8.0.0] 2020-07-01

### Updates from upstream (FTW-daily v6.0.0)

- [change] Use privileged transitions for price calculation by default and update the process alias.
  [#1314](https://github.com/sharetribe/ftw-daily/pull/1314)
- [add] Add client secret enquiry to 'yarn run config' script
  [#1313](https://github.com/sharetribe/ftw-daily/pull/1313)
- [change] Add UI support for flexible pricing and privileged transitions. Note that this requires
  updating the booking breakdown estimation code that is now done in the backend.
  [#1310](https://github.com/sharetribe/ftw-daily/pull/1310)
- [add] Add local API endpoints for flexible pricing and privileged transitions
  [#1301](https://github.com/sharetribe/ftw-daily/pull/1301)
- [fix] `yarn run dev-backend` was expecting NODE_ENV.
  [#1303](https://github.com/sharetribe/ftw-daily/pull/1303)

[v8.0.0]: https://github.com/sharetribe/ftw-hourly/compare/v7.0.0...v8.0.0

## [v7.0.0] 2020-06-04

### Updates from upstream (FTW-daily v5.0.0)

- [change] Streamlining filter setup. Everyone who customizes FTW-templates, needs to update filters
  and unfortunately the related code has been spread out in multiple UI containers.

  Now, filters are more configurable through marketplace-custom-config.js. You can just add new
  filter configs to `filters` array in there - and that should be enough for creating new filters
  for extended data.

  If your are creating a totally new filter component, you can take it into use in a single file:
  src/containers/SearchPage/FilterComponent.js

  In addition, we have renamed couple of container components:

  - SearchFilters -> SearchFiltersPrimary
  - SearchFiltersPanel -> SearchFiltersSecondary (SearchFiltersMobile has kept its name.)

  SortBy filter's state is also tracked similarly as filters. From now on, the state is kept in
  MainPanel and not in those 3 different UI containers.

  [#1296](https://github.com/sharetribe/ftw-daily/pull/1296)

[v7.0.0]: https://github.com/sharetribe/ftw-hourly/compare/v6.6.0...v7.0.0

## [v6.6.0] 2020-06-04

### Updates from upstream

- [fix] In some situations, ProfileMenu has began to overflow on TopbarDesktop.
  [#1290](https://github.com/sharetribe/ftw-daily/pull/1290)
- [change] Update dependencies (patch updates only)
  [#1291](https://github.com/sharetribe/ftw-daily/pull/1291)
- [change] Refactor server API routes into separate files.
  [#1294](https://github.com/sharetribe/ftw-daily/pull/1294)
- [change] Start the backend API router in dev mode with a dev server.
  [#1297](https://github.com/sharetribe/ftw-daily/pull/1297)

[v6.6.0]: https://github.com/sharetribe/ftw-hourly/compare/v6.5.1...v6.6.0

## [v6.5.1] 2020-05-13

- [fix] Check length of `selectedConfigOptions` in `SectionFeaturesMaybe` to choose between one and
  two column layout. [#92](https://github.com/sharetribe/ftw-hourly/pull/90)

### Updates from upstream

This is update from [upstream](https://github.com/sharetribe/ftw-daily): v4.4.3

- [fix] Allow white space on Japanese bank account info. Japan collects bank name and account owner
  name in addition to routing numbers. [#1287](https://github.com/sharetribe/ftw-daily/pull/1287)
- [fix] wrongly named default props handleSubmit renamed to onSubmit
  [#1288](https://github.com/sharetribe/ftw-daily/pull/1288)

## [v6.5.0] 2020-04-17

- [change] Reorganize search filters [#89](https://github.com/sharetribe/ftw-hourly/pull/89)
- [fix] Remove unnecessary required props, which were introduced accidentally in upstream update
  (v6.3.0). [#88](https://github.com/sharetribe/ftw-hourly/pull/88/)
- [add] Time-based availability filtering with a start/end date + min booking duration search filter
  [#85](https://github.com/sharetribe/ftw-hourly/pull/85)

## [v6.4.1] 2020-04-09

### One change in this template


src/config.js
import * as custom from './marketplace-custom-config.js';
import defaultLocationSearches from './default-location-searches';
import { defaultMCC, stripePublishableKey, stripeCountryDetails } from './stripe-config';
import { currencyConfiguration } from './currency-config';

const env = process.env.REACT_APP_ENV;
const dev = process.env.REACT_APP_ENV === 'development';

// If you want to change the language, remember to also change the
// locale data and the messages in the app.js file.
const locale = 'en';
const i18n = {
  /*
    0: Sunday
    1: Monday
    ...
    6: Saturday
  */
  firstDayOfWeek: 0,
};

// Should search results be ordered by distance to origin.
// NOTE: If this is set to true add parameter 'origin' to every location in default-location-searches.js
// Without the 'origin' parameter, search will not work correctly
// NOTE: Keyword search and ordering search results by distance can't be used at the same time. You can turn keyword
// search off by changing the keywordFilterConfig parameter active to false in marketplace-custom-config.js
const sortSearchByDistance = false;

// API supports custom processes to be used in booking process.
// We need to specify it when we are initiating a new order
// (or fetching price information by calling 'initiateSpeculative' endpoint).
//
// In a way, 'processAlias' defines which transaction process (or processes)
// this particular web application is able to handle.
const bookingProcessAlias = 'flex-hourly-default-process/release-1';

// The transaction line item code for the main unit type in bookings.
//
// Possible values: ['line-item/night', 'line-item/day', 'line-item/units';]
//
// Note 1: This 'bookingUnitType' variable affects only web app.
//         If you are using privileged transitions (which is used by the default process),
//         you also need to configure unit type in API server: server/api-util/lineItems.js
//
// Note 2: Translations will use different translation keys for night, day or unit
//         depending on the value chosen.
const bookingUnitType = 'line-item/units';

// Should the application fetch available time slots (currently defined as
// start and end dates) to be shown on listing page.
const enableAvailability = process.env.REACT_APP_AVAILABILITY_ENABLED === 'true';

// A maximum number of days forwards during which a booking can be made.
// This is limited due to Stripe holding funds up to 90 days from the
// moment they are charged. Also note that available time slots can only
// be fetched for 180 days in the future.
const dayCountAvailableForBooking = 90;

// To pass environment variables to the client app in the build
// script, react-scripts (and the sharetribe-scripts fork of
// react-scripts) require using the REACT_APP_ prefix to avoid
// exposing server secrets to the client side.
const sdkClientId = process.env.REACT_APP_SHARETRIBE_SDK_CLIENT_ID;
const sdkBaseUrl = process.env.REACT_APP_SHARETRIBE_SDK_BASE_URL;
const sdkTransitVerbose = process.env.REACT_APP_SHARETRIBE_SDK_TRANSIT_VERBOSE === 'true';

const currency = process.env.REACT_APP_SHARETRIBE_MARKETPLACE_CURRENCY;

// Currency formatting options.
// See: https://github.com/yahoo/react-intl/wiki/API#formatnumber
const currencyConfig = currencyConfiguration(currency);

// Listing minimum price in currency sub units, e.g. cents.
// 0 means no restriction to the price
const listingMinimumPriceSubUnits = 0;

// Sentry DSN (Data Source Name), a client key for authenticating calls to Sentry
const sentryDsn = process.env.REACT_APP_SENTRY_DSN;

// If webapp is using SSL (i.e. it's behind 'https' protocol)
const usingSSL = process.env.REACT_APP_SHARETRIBE_USING_SSL === 'true';

// Address information is used in SEO schema for Organization (http://schema.org/PostalAddress)
const addressCountry = 'FI';
const addressRegion = 'Helsinki';
const postalCode = '00100';
const streetAddress = 'Bulevardi 14';

// Canonical root url is needed in social media sharing and SEO optimization purposes.
const canonicalRootURL = process.env.REACT_APP_CANONICAL_ROOT_URL;

// Site title is needed in meta tags (bots and social media sharing reads those)
const siteTitle = 'TalentBookr';

// Twitter handle is needed in meta tags (twitter:site). Start it with '@' character
const siteTwitterHandle = '@sharetribe';

// Instagram page is used in SEO schema (http://schema.org/Organization)
const siteInstagramPage = null;

// Facebook page is used in SEO schema (http://schema.org/Organization)
const siteFacebookPage = 'https://www.facebook.com/Sharetribe/';

// Facebook counts shares with app or page associated by this id
// Currently it is unset, but you can read more about fb:app_id from
// https://developers.facebook.com/docs/sharing/webmasters#basic
// You should create one to track social sharing in Facebook
const facebookAppId = null;

const maps = {
  mapboxAccessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
  googleMapsAPIKey: process.env.REACT_APP_GOOGLE_MAPS_API_KEY,

  // The location search input can be configured to show default
  // searches when the user focuses on the input and hasn't yet typed
  // anything. This reduces typing and avoids too many Geolocation API
  // calls for common searches.
  search: {
    // When enabled, the first suggestion is "Current location" that
    // uses the browser Geolocation API to query the user's current
    // location.
    suggestCurrentLocation: process.env.REACT_APP_DEFAULT_SEARCHES_ENABLED === 'true',

    // Distance in meters for calculating the bounding box around the
    // current location.
    currentLocationBoundsDistance: 1000,

    // Example location can be edited in the
    // `default-location-searches.js` file.
    defaults:
      process.env.REACT_APP_DEFAULT_SEARCHES_ENABLED === 'true' ? defaultLocationSearches : [],

    // Limit location autocomplete to a one or more countries
    // using ISO 3166 alpha 2 country codes separated by commas.
    // If you want to limit the autocomplete, uncomment this value:
    // countryLimit: ['AU'],
  },

  // When fuzzy locations are enabled, coordinates on maps are
  // obfuscated randomly around the actual location.
  //
  // NOTE: This only hides the locations in the UI level, the actual
  // coordinates are still accessible in the HTTP requests and the
  // Redux store.
  fuzzy: {
    enabled: false,

    // Amount of maximum offset in meters that is applied to obfuscate
    // the original coordinates. The actual value is random, but the
    // obfuscated coordinates are withing a circle that has the same
    // radius as the offset.
    offset: 500,

    // Default zoom level when showing a single circle on a Map. Should
    // be small enough so the whole circle fits in.
    defaultZoomLevel: 13,

    // Color of the circle on the Map component.
    circleColor: '#c0392b',
  },

  // Custom marker image to use in the Map component.
  //
  // NOTE: Not used if fuzzy locations are enabled.
  customMarker: {
    enabled: false,

    // Publicly accessible URL for the custom marker image.
    //
    // The easiest place is /public/static/icons/ folder, but then the
    // marker image is not available while developing through
    // localhost.
    url: encodeURI(`${canonicalRootURL}/static/icons/map-marker-32x32.png`),

    // Dimensions of the marker image.
    width: 32,
    height: 32,

    // Position to anchor the image in relation to the coordinates,
    // ignored when using Mapbox.
    anchorX: 16,
    anchorY: 32,
  },
};

// NOTE: only expose configuration that should be visible in the
// client side, don't add any server secrets in this file.
const config = {
  env,
  dev,
  locale,
  bookingProcessAlias,
  bookingUnitType,
  enableAvailability,
  dayCountAvailableForBooking,
  i18n,
  sdk: {
    clientId: sdkClientId,
    baseUrl: sdkBaseUrl,
    transitVerbose: sdkTransitVerbose,
  },
  sortSearchByDistance,
  currency,
  currencyConfig,
  listingMinimumPriceSubUnits,
  stripe: {
    defaultMCC: defaultMCC,
    publishableKey: stripePublishableKey,
    supportedCountries: stripeCountryDetails,
  },
  canonicalRootURL,
  address: {
    addressCountry,
    addressRegion,
    postalCode,
    streetAddress,
  },
  siteTitle,
  siteFacebookPage,
  siteInstagramPage,
  siteTwitterHandle,
  facebookAppId,
  sentryDsn,
  usingSSL,
  maps,
  custom,
};

export default config;
